farm/data_handler/processor.py

```diff
@@ -1727,6 +1728,292 @@ def _sample_to_features(self, sample) -> dict:
+        config = json.load(open(processor_config_file))
+        # init tokenizer
+        tokenizer = Tokenizer.load(load_dir, tokenizer_class=config["tokenizer"])
+        passage_tokenizer = Tokenizer.load(load_dir, tokenizer_class=config["passage_tokenizer"])
+
+        # we have to delete the tokenizer string from config, because we pass it as Object
+        del config["tokenizer"]
+        del config["passage_tokenizer"]
+
+        processor = cls.load(tokenizer=tokenizer, passage_tokenizer=passage_tokenizer, processor_name="TextSimilarityProcessor", **config)
+        for task_name, task in config["tasks"].items():
+            processor.add_task(name=task_name, metric=task["metric"], label_list=task["label_list"])
+
+        if processor is None:
+            raise Exception
+
+        return processor
+
+    def save(self, save_dir):
+        """
+        Saves the vocabulary to file and also creates a json file containing all the
+        information needed to load the same processor.
+
+        :param save_dir: Directory where the files are to be saved
+        :type save_dir: str
+        """
+        os.makedirs(save_dir, exist_ok=True)
+        config = self.generate_config()
+        # save tokenizer incl. attributes
+        config["tokenizer"] = self.tokenizer.__class__.__name__
+        config["passage_tokenizer"] = self.tokenizer.__class__.__name__
+
+        # Because the fast tokenizers expect a str and not Path
+        # always convert Path to str here.
+        self.tokenizer.save_pretrained(str(save_dir))
+        self.passage_tokenizer.save_pretrained(str(save_dir))
+
+        # save processor
+        config["processor"] = self.__class__.__name__
+        output_config_file = Path(save_dir) / "processor_config.json"
+        with open(output_config_file, "w") as file:
+            json.dump(config, file)
+
+    def file_to_dicts(self, file: str) -> [dict]:
+        """
+        Converts a Dense Passage Retrieval (DPR) data file in json format to a list of dictionaries.
+
+        :param file: filename of DPR data in json format
+
+        Returns:
+        list of dictionaries: List[dict]
+        each dictionary:
+        {"query": str,
+        "passages": [{"text": document_text, "title": xxx, "label": "positive", "external_id": abb123},
+        {"text": document_text, "title": xxx, "label": "hard_negative", "external_id": abb134},
+        ...]}
+        """
+        dicts = read_dpr_json(file)
+        return dicts
+
+    def _normalize_question(self, question: str) -> str:
+        """
+        Removes '?' from queries/questions
+
+        :param question: string representing the question
+
+        Returns:
+            Question without the '?'
+        """
+        if question[-1] == '?':
+            question = question[:-1]
+        return question
+
+    def _dict_to_samples(self, dictionary: dict, **kwargs) -> [Sample]:
+        """
+        dictionary representing 1 sample to Sample for DPR
```

Comment:
Change Docstring to "Creates one sample from one dict consisting of the query, positive passages and hard negative passages"